### PR TITLE
Fix join when the list only contains 1 element

### PR DESCRIPTION
**Pull request by zmingxie**
This is for https://github.com/bazaarvoice/cloudformation-ruby-dsl/issues/58

### CODE REVIEW

File: lib/cloudformation-ruby-dsl/dsl.rb [hunk 1]
```diff
@@ -151,7 +151,7 @@ def get_azs(region = '') { :'Fn::GetAZs' => region } end
 def join(delim, *list)
   case list.length
     when 0 then ''
-    when 1 then list[0]
+    when 1 then join_list(delim, list[0])
```
Comment (jonaf) [inline lib/cloudformation-ruby-dsl/dsl.rb]: Actually, I think we should remove this line altogether and always return `join_list(delim,list)` even for a list of size 1. This way, the second parameter of the `Fn::Join` intrinsic function always produces a list, even if it's of size 1, and never an object.

The current code will produce something like:

```json
"SecurityGroups": [
          {
            "Fn::Join": [
              ",",
              {
                "Ref": "SecurityGroups"
              }
            ]
          }
        ]
```

However, according to the CloudFormation documentation, we should always yield a list. Thus, this is the expected output:

```json
        "SecurityGroups": [
          {
            "Fn::Join": [
              ",",
              [
                {
                  "Ref": "SecurityGroups"
                }
              ]
            ]
          }
        ]
```

Would you mind just removing line 154 entirely and seeing how that works for you?

Comment (zmingxie) [inline lib/cloudformation-ruby-dsl/dsl.rb]: @jonaf Your suggestion was my first attempt as well. However, I got the following error when I put only 1 element in the list:
```
Template validation error: Template error: every Fn::Join object requires two parameters, (1) a string delimiter and (2) a list of strings to be joined or a function that returns a list of strings (such as Fn::GetAZs) to be joined.
```
That's why I kept this 1 element as a special case.

Comment (jonaf) [inline lib/cloudformation-ruby-dsl/dsl.rb]: What did the expanded template look like when you got that error? I was not able to reproduce it. Removing this line should produce a delimiter and a list, of size 1, which should be valid.

Comment (zmingxie) [inline lib/cloudformation-ruby-dsl/dsl.rb]: the expanded template looks exactly like your example above. Here is the diff:
```
$ diff original.json mod.json
254,256c254,258
<               {
<                 "Ref": "SecurityGroupIds"
<               }
---
>               [
>                 {
>                   "Ref": "SecurityGroupIds"
>                 }
>               ]
```
I agree with you this should be valid, and I really don't know why I'm getting this error...

Comment (jonaf) [inline lib/cloudformation-ruby-dsl/dsl.rb]: Are you sure you're not getting your tests confused? I don't get the validation error when Line 154 is *entirely removed*, but when I change it to `when 1 then join_list(delim,list[0])`, I get the validation error you mentioned before.

Comment (zmingxie) [inline lib/cloudformation-ruby-dsl/dsl.rb]: Nope, I just double checked. How many elements do you have in your "SecurityGroupIds"? For my test, I only have 1 SG in the list.

Comment (jonaf) [inline lib/cloudformation-ruby-dsl/dsl.rb]: Here's my sample DSL:

```ruby
  resource 'TestSecurityGroup', :Type => 'AWS::EC2::SecurityGroup', :Properties => {
      :GroupDescription     => 'Enable all traffic',
      :SecurityGroupIngress => [
          {:IpProtocol => '-1', :FromPort => '-1', :ToPort => '-1', :CidrIp => '0.0.0.0'}
      ],
  }

  output 'TestOutput', {
    :Value => join(',', ref('TestSecurityGroup'))
  }
```

When I expand it, I get the following:

```json
"Outputs": {
    "TestOutput": {
      "Value": {
        "Fn::Join": [
          ",",
          [
            {
              "Ref": "TestSecurityGroup"
            }
          ]
        ]
      }
    }
  }
```

When I run the `validate` subcommand, I get the expected message, "Validation successful."

Comment (zmingxie) [inline lib/cloudformation-ruby-dsl/dsl.rb]: Oh... I think I know why, in your test, the type for 'TestSecurityGroup' is 'AWS::EC2::SecurityGroup'. My test defined 'SecurityGroupIds' as 'List<AWS::EC2::SecurityGroup::Id>' which itself is a list. That causes the confusions... Let me update my patch and remove this special case.

**Comment by zmingxie** [comment]
Patch is updated.

**Comment by jonaf** [comment]
Merged. Thanks!

**Pull request closed without merging**